ext/MitsubaLoader/CMitsubaLoader.h

```diff
@@ -16,6 +16,304 @@ namespace ext
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_SHIFT_PHONG_EXP_TEX = 16u;
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_MASK_MIXTURE_CHILD_COUNT = 0x1fu;
+	_IRR_STATIC_INLINE_CONSTEXPR uint32_t BITFIELDS_SHIFT_MIXTURE_CHILD_COUNT = 10u;
+#include "irr/irrpack.h"
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec2 alpha;
+		vec3 reflectance;
+	};
+	*/
+	struct alignas(16) SAllDiffuse
+	{
+		uint32_t bitfields;
+		float dummy;
+		//if flag decides to use alpha texture, {alpha[0..1]} is bindless texture ID (bit-cast to uvec2)
+		//otherwise alpha[0] is single-float alpha
+		float alpha[2];
+		//if flag decides to use reflectance texture, {reflectance[0..1]} is bindless texture ID (bit-cast to uvec2)
+		//otherwise reflectance[0..2] is constant reflectance
+		float reflectance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec3 transmittance;
+	};
+	*/
+	struct alignas(16) SDiffuseTransmitter
+	{
+		uint32_t bitfields;
+		float dummy[3];
+		//if flag decides to use transmittance texture, {transmittance[0..1]} is bindless texture ID (bit-cast to uvec2)
+		//otherwise transmittance[0..2] is constant transmittance
+		float transmittance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		vec2 int_ext_ior;
+		vec2 alpha_u;
+		vec2 alpha_v;
+		vec3 specReflectance;
+		vec3 specTransmittance;
+	}
+	*/
+	struct alignas(16) SAllDielectric
+	{
+		uint32_t bitfields;
+		float dummy0;
+		//int_ext_ior[0] is internal IoR, int_ext_ior[1] is external IoR
+		float int_ext_ior[2];
+		//if NDF is Ashikhmin-Shirley:
+		//	if flag decides to use alpha_u texture, alpha_u[0..1] is bindless texture ID
+		//	otherwise alpha_u[0] is constant alpha_u
+		//	if flag decides to use alpha_v texture, alpha_v[0..1] is bindless texture ID
+		//	otherwise alpha_v[0] is constant alpha_v
+		//otherwise (different NDF)
+		//	if flag decides to use alpha texture, alpha_u[0..1] is bindless texture ID
+		//	otherwise alpha_u[0] is constant alpha
+		float alpha_u[2];
+		float alpha_v[2];
+		//if flag decides to use spec refl texture, specReflectance[0..1] is bindless texture ID
+		//otherwise specReflectance[0..2] is constant spec refl
+		float specReflectance[3];
+		float dummy1;
+		//if flag decides to use spec transmitance texture, specTransmittance[0..1] is bindless texture ID
+		//otherwise specTransmittance[0..2] is constant spec transmitance
+		float specTransmittance[3];
+	} PACK_STRUCT;
+	/**
+	GLSL:
+	struct {
+		uint bitfields;
+		float ext_ior;
+		vec2 alpha_u;
+		vec2 alpha_v;
+		layout (column_major) mat2x3 ior;
+		vec3 specReflectance;
+	};
+	*/
+	struct alignas(16) SAllConductor
+	{
+		uint32_t bitfields;
+		float ext_ior;
+		//same as for SAllDielectric::alpha_u,alpha_v
+		float alpha_u[2];
+		float alpha_v[2];
+		float dummy[2];
+		//column-major matrix 3x2 (GLSL's 2x3)
+		//ior[0..2] is column 0 (real part of IoR)
+		//ior[4..6] is column 1 (imaginary part of IoR)
+		float ior[8];
```

Comment:
Precompute a single `eta` value (factor out the external ior)

you could store them in the yzw components instead of xyz, so then the first 4 bytes could be the `bitfields` (then put your `alpha` parameters last)

Reply:
actually discord convo supersedes all "member layout suggestions"